urlfc1c29991b_nil_mish_InJava_tgz-pJ8-edu_usc_cs_nsl_indexing_test_TestTrieProtoJ8/src/edu/usc/cs/nsl/counting/CountingUtil.java

```diff
@@ -0,0 +1,494 @@
+		BufferedWriter bw = new BufferedWriter(new FileWriter(tmpDir+"/chunk_list_"+max_string_size+".txt"));
+		bw.write('0');
+		for(int i = 0; i < big_array_size; i++){
+			if(the_big_array[i] > 0){
+				number_of_words_in_chunk += the_big_array[i]; 
+				this_chunk_size++;
+				if(this_chunk_size == chunk_size - 1){
+					this_chunk_size = 0;
+					num_of_chunks++;
+					bw.write("\t"+(i+1)+"\t"+number_of_words_in_chunk+"\n"+(i+1)); // start of range is inclusive end is exclusive.
+					number_of_words_in_chunk = 0;
+				}
+			}
+		}
+		num_of_chunks++;
+		bw.write("\t"+big_array_size+"\t"+number_of_words_in_chunk+"\n");
+		bw.close();
+		System.out.println("Max String: "+max_string_size+"\tNumLines: "+num_lines+"\tnum of chunks: "+num_of_chunks);
+		return num_of_chunks;
+	}
+	
+	
+	/**
+	 * Function to process chunks. This function uses hashing to do counting.
+	 * Count for words without hash index collision are written to final 
+	 * output file. Words with index collision have the indexes written into 
+	 * a temporary file.
+	 * Note: There are collisions in the RAM counting phase (same big index) 
+	 * and during the processing of chunks (this index). Both these cases
+	 * need to be handled separately.
+	 * 
+	 * @param max_string_size
+	 * @param num_chunks
+	 * @param output_file
+	 * @throws IOException
+	 */
+	
+	public void processChunks(int max_string_size, int num_chunks,
+			String output_file) throws IOException {
+		// The chunks can be loaded in memory and hence they can be counted.
+		int chunk_size = closestPrime(max_ram/(max_string_size + 8));
+		System.out.println("Big array size: "+big_array_size+"\t chunk size: "+chunk_size);
+		BufferedReader chunk_list_file = null;
+		chunk_list_file = new BufferedReader(new FileReader(getTmpChunkListFileName(max_string_size)));
+		String the_line = chunk_list_file.readLine();		
+		int chunk_index[][] = new int[num_chunks][2];
+		int index = 0;
+		int num_collision = 0;
+		while(the_line != null){
+			the_line.replace("\n","").replace("\r", "");
+			if(the_line.equals("")){
+				the_line = chunk_list_file.readLine();
+				continue;
+			}
+			String tokens[] = the_line.split("\t");
+			if(tokens[0] != null && tokens[1] != null){
+				chunk_index[index][0] = Integer.parseInt(tokens[0]);
+				chunk_index[index][1] = Integer.parseInt(tokens[1]);
+			}
+			index++;
+			the_line = chunk_list_file.readLine();
+		}
+		chunk_list_file.close();
+		
+		BufferedWriter bw_collision = new BufferedWriter(new FileWriter(getCollisionTmpFileName(max_string_size),true));
+
+		for (int this_chunk_index = 0; this_chunk_index < chunk_index.length; this_chunk_index++) {			
+			String word_array[] = new String[chunk_size];
+			long count_array[] = new long[chunk_size];
+			int start_index = chunk_index[this_chunk_index][0];
+			int end_index = chunk_index[this_chunk_index][1];
+			BufferedReader br = null;
+			br = new BufferedReader(new FileReader(getTmpFileName(max_string_size)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *br*. Currently, there are execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).